src/transfers.ts

```diff
@@ -0,0 +1,176 @@
+                             findInstruction(message, 'ReceiveTeleportedAsset');
+    
+    if (!assetInstruction) {
+        console.error(`XCM. No assets instructions found. Block: ${blockNumber}`);
+        return result
+    }
+    // Destination chain and tokens info are not yet used.
+    // Ex.: const destChain = getLocation(destination, chainInfo, blockNumber)
+    const transfers = getTokenAmountFromAsset(
+        assetInstruction,
+        chainInfo,
+        blockNumber
+    )
+    if (transfers.length > 0) {
+        const [_token, amount] = transfers[0]
+        result.amount = amount
+    } else {
+        console.error(`XCM. No assets found inside of instruction. Block: ${blockNumber}`);
+    }
+    return result
+}
+
+function getLocation(location: StagingXcmV4Location, chainInfo: ChainInfo, blockNumber: number): string {
+    if (!(location.interior.isX1)) {
+        log.error(`XCM. Junctions not supported: ${location.interior.type}. Block ${blockNumber}`)
+        return 'Unknown'
+    }
+    const x1 = location.interior.asX1[0]
+    if (x1.isAccountId32) {
+        const originIdHex = x1.asAccountId32.id.toString();
+        return encodeAddress(hexToU8a(originIdHex), chainInfo.SS58);
+    } else if (x1.isAccountKey20) {
+        return x1.asAccountKey20.key.toHuman()
+    } else if (x1.isParachain) {
+        const chainIndex =  x1.asParachain.toString()
+        return parachainNames[chainInfo.id][chainIndex] || `Parachain ${chainIndex}`;
+    } else {
+        log.error(`XCM. Junctions not supported: ${x1.type}. Block ${blockNumber}`)
+    }
+    return 'Unknown'
+}
+
+const findInstruction = (xcm: StagingXcmV4Xcm, key: string) => xcm.find(instr => instr.type === key)?.['as' + key];
+
+function getTokenAmountFromAsset(assets: StagingXcmV4Asset[], chainInfo: ChainInfo, blockNumber: number): [string, string][] {
+    const result: [string, string][] = [];
+    for (const asset of assets) {
+        if (!(asset.fun.isFungible)) {
+            // NFTs are not a subject for monitoring
+            continue
+        }
+        const amount = formatBalance(asset.fun.asFungible, {decimals: chainInfo.decimals[0], withSi: false, forceUnit: '-' })
+        const interior = asset.id.interior
+        let token: string;
+        if (interior.isHere) {
+            token = chainInfo.tokens[0]
+        } else if (interior.isX3) {
+            // Not supported yet. Should be processed for the parachain monitoring
+            continue
+        } else {
+            log.error(`Asset Junctions not supported: ${interior.type}. Block ${blockNumber}`)
+            continue
+        }
+        result.push([token, amount])
+    }
+    return result
+}
+
+export const isBalancesTransferEvent = (event: Event): boolean => {
+    return event.section === 'balances' && event.method === 'Transfer'
+}
+
+export function isXcmSentEvent(event: Event): boolean {
+    return ['polkadotXcm', 'xcmPallet'].includes(event.section) && event.method === 'Sent'
```

Comment:
do we need kusamaXcm maybe ?

Reply:
`xcmPallet` - relay chain monitoring.
`polkadotXcm` - AssetHub monitoring for both Polkadot and Kusama (we don't use it now).

And as for the parachain monitoring, it should be checked individually. For example Centrifuge uses `xtokens` pallet and `TransferredMultiAssets` event. The processing is different. Example: https://centrifuge.subscan.io/event?extrinsic=4551454-2